rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -342,9 +342,16 @@ struct user_callback_data_t
   rmw_event_callback_t callback {nullptr};
   const void * user_data {nullptr};
   size_t unread_count {0};
-  rmw_event_callback_t event_callback[DDS_STATUS_ID_MAX + 1] {nullptr};
-  const void * event_data[DDS_STATUS_ID_MAX + 1] {nullptr};
-  size_t event_unread_count[DDS_STATUS_ID_MAX + 1] {0};
+  rmw_event_callback_t event_callback[RMW_EVENT_INVALID + 1] {nullptr};
+  const void * event_data[RMW_EVENT_INVALID + 1] {nullptr};
+  size_t event_unread_count[RMW_EVENT_INVALID + 1] {0};
```

Comment:
> these changes are required because there are two different events for matched and unmatched, right? (https://github.com/ros2/rmw_fastrtps/pull/645#discussion_r1034772808)

Yes. Original code is based on `DDS_STATUS_ID` which only matched event.  
So in order to support unmatched event, I changed it to use RMW_EVENT ID.   

> I understand that different events can avoid user to do the further work to see which event comes in, but also we are providing the almost same event in the implementation...  

rmw_take_event() will take event status from last read.   
Now we use 2 RMW events for one DDS event, so we have to save the status of DDS event for each RMW event.   

I discussed with Chen-san. Now we only know one real case. While first matched event (for subscription or publisher) happen, start certain service. While the last subscription or publisher is disconnected, stop certain service.   

For this case, user will do based on matched/unmatched event
- While RMW_EVENT_PUBLICATION_MATCHED/RMW_EVENT_SUBSCRIPTION_MATCHED is triggered, user check service is started. If not, start this service. Or call rmw_take_event() to check if current_matched_count is 1 and start service if count is 1. 
- While RMW_EVENT_PUBLICATION_UNMATCHED/RMW_EVENT_SUBSCRIPTION_UNMATCHED is triggered, user need to call rmw_take_event() to check if current_matched_count is 0. If count is 0, stop service.  

If only one matched event (Same as DDS), user do the same thing as above.   
Using 2 events doesn't bring convenience for user.  

If user only care matched and unmatched event, current 2 events implementation can bring convenience.  
2 events implementation introduce extra logic to rmw layer.  I think https://github.com/ros2/rmw_fastrtps/pull/645#discussion_r1034772808 don't want to have extra logic in rmw layer.  
I have no enough requirement from real cases. Which implementation (2 events or 1 event) is better one.

Reply:
> I have no enough requirement from real cases.

That makes two of us ...

I would even go further: if the only foreseen use is starting/stopping a service, you could also simply do that based on a single event that merely says something changed in the set of matched readers. You use that event as a trigger, and then check that discovery information/graph cache to see whether you just went from 0 to 1 or from 1 to 0.

At least in this RMW layer, a quick reading of the changes suggests the code (almost) duplication could be eliminated by adding a few functions with well-chosen parameters.

Reply:
@eboasson thanks for the comment!
CC: @Barry-Xu-2018 

yeah, that is correct. i mean, we can just generate the event that tells us `something` happened on either publisher or subscription, then we can internally calls graph interfaces to return the number of endpoints. I would think most likely that works for many use cases.

one thing that i want to hear is there will be time window between triggering event from implementation and getting the graph information. these cannot be synced, so there might be event but it does not show any difference the graph change? this cannot be supported in current interfaces w/o this one?

Reply:
> these cannot be synced, so there might be event but it does not show any difference the graph change? this cannot be supported in current interfaces w/o this one?

I would think so. If you need that information, then you do need something like this.